### ISSUE TITLE
menu_get_item() doesnt return a node object

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#944]
This may not be a bug. I'm porting [token filter](https://www.drupal.org/project/token_filter) and it works, but only for global tokens. This module uses, strangely, a call to `debug_backtrace()` to trace when `field_default_view()` is called to get the node object context. That, apparently seemed the best way to get the node context from `hook_filter_info()`. 

But `field_default_view()` isnt called in Backdrop it seems.

There was an alternative suggestion to [use menu_get_item instead of debug_backtrace to determine context](https://www.drupal.org/node/2144145) so I tried that too, but `menu_get_item()` doesnt return a node object in Backdrop.

Is this a bug?

If not can anyone figure out any way to get the node context from `hook_lifter_info()` in Backdrop?

I also tried `layout_load('default)` then `getContext()` but ony the user context is available there. And [finding the current active layout is not possible](https://github.com/backdrop/backdrop-issues/issues/791).

I'm all tied up.

Here is all of token_filter for reference:

```
/**
 * Implements hook_filter_info().
 */
function token_filter_filter_info() {
  $filters['filter_tokens'] = array(
    'title' => t('Replace tokens'),
    'description' => t('The usage of this filter should be restricted to trusted users only as tokens with sensitive data could be exposed.'),
    'process callback' => '_token_filter_filter_tokens',
    'tips callback' => '_token_filter_filter_tips',
    'cache' => FALSE,
  );
  return $filters;
}

/**
 * Filter process callback for the token text filter.
 */
function _token_filter_filter_tokens($text, $filter, $format, $langcode, $cache, $cache_id) {
  $data = array();
  $options = array();

  // Attempt to figure out the current context based on the current backtrace.
  $backtrace = debug_backtrace();
  array_shift($backtrace); // Pop off this current function in the stack.
  foreach ($backtrace as $caller) {
    switch ($caller['function']) {
      case 'field_default_view':
        // field_default_view($entity_type, $entity) is fairly reliable since
        // it is called by both field_attach_view() and field_view_field().
        $entity_type = $caller['args'][0];
        $entity = $caller['args'][1];
        $token_type = token_get_entity_mapping('entity', $entity_type);
        $data[$token_type] = $entity;
        // Use the proper language code that field_default_view() was called with.
        if ($langcode = $caller['args'][4]) {
          $language_list = language_list();
          if (!empty($language_list[$langcode])) {
            $options['language'] = $language_list[$langcode];
          }
        }
        break;
      case '_drupal_bootstrap_full':
      case 'menu_execute_active_handler':
        break 2;
    }
  }

  return token_replace($text, $data, $options);
}
```

**Comment by quicksketch** [comment]
Wow, well I'm not sure what to recommend here. I don't think menu_get_item() would help you here because it's only going to return a node when you're on the node page itself (node/x). If you viewed a node's teaser, such as on the homepage, then menu_get_item() won't return anything. That's probably why debug_backtrace() is being used. I think your only option in porting is to work through the new backtrace to figure out when this filter is being called.

**Comment by docwilmot** [comment]
This cannot be done it seems. I've tried everything. In a list of nodes, there is no way to get the node context from hook_filter_info().

**Comment by quicksketch** [comment]
Hi @docwilmot, I put a dsm(debug_backtrace()) in a filter callback and it looks like the current approach would still work. `field_attach_load()` is indeed still called, and it still has the Node object as a parameter:

![backtrace](https://cloud.githubusercontent.com/assets/100206/7763451/f9875b12-fffd-11e4-8804-de8c65ac0e42.png)

The code may need to be adjusted slightly, but the node is definitely still available in the same way (as hacky as it may be).

**Comment by quicksketch** [comment]
Oh, I suppose previously it was `field_default_view()` not `field_attach_load()`. But in any case, you should be able to get the Node (or other entity object) if needed.

In Backdrop I've found that the easiest way of identifying a Node or Entity is to use the PHP `is_a()` function to identify entities. e.g.

```php
if (is_a($args[1], 'Entity')) {
  debug($entity->label());
}
```

**Comment by quicksketch** [comment]
Oops, didn't mean to reopen. :)

**Comment by docwilmot** [comment]
Thanks, I had tried `field_attach_load` yesterday, but it includes every node object on the page, (i.e. `$args[1]` isnt just a single node like in your example) so its not possible to tell which exact node we're dealing with. I tried `reset($args[1])` but that is unreliable. And it also doesnt help with taxonomy terms. Pretty sure I've thoroughly examined every possible function in that backtrace and come up short. 

I'm figuring that this token-in-fields functionality may need to be done by directly calling token_replace on field content rather than through filters. Like [this Advanced Text Formatter](http://drupal.org/project/advanced_text_formatter) does.
